vm/devices/storage/storage_tests/tests/scsidvd_nvme.rs

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#![cfg(target_os = "linux")]
+
+use chipset_device::mmio::ExternallyManagedMmioIntercepts;
+use disk_backend::Disk;
+use disk_nvme::NvmeDisk;
+use guestmem::GuestMemory;
+use guid::Guid;
+use nvme::NvmeController;
+use nvme::NvmeControllerCaps;
+use nvme_driver::NvmeDriver;
+use pal_async::async_test;
+use pal_async::DefaultDriver;
+use pci_core::msi::MsiInterruptSet;
+use scsi_buffers::OwnedRequestBuffers;
+use scsi_buffers::RequestBuffers;
+use scsi_core::AsyncScsiDisk;
+use scsi_core::Request;
+use scsi_defs::ScsiOp;
+use scsi_defs::ISO_SECTOR_SIZE;
+use scsidisk::scsidvd::SimpleScsiDvd;
+use test_with_tracing::test;
+use user_driver::emulated::DeviceSharedMemory;
+use user_driver::emulated::EmulatedDevice;
+use vmcore::vm_task::SingleDriverBackend;
+use vmcore::vm_task::VmTaskDriverSource;
+use zerocopy::IntoBytes;
+
+struct ScsiDvdNvmeTest {
+    scsi_dvd: SimpleScsiDvd,
+    _nvme_driver: NvmeDriver<EmulatedDevice<NvmeController>>, // We need to store this to keep it from going out of scope
+}
+
+impl ScsiDvdNvmeTest {
+    async fn new(
+        driver: DefaultDriver,
+        sector_size: u32,
+        sector_count: u64,
+        read_only: bool,
+    ) -> Self {
+        const MSIX_COUNT: u16 = 2;
+        const IO_QUEUE_COUNT: u16 = 64;
+        const CPU_COUNT: u32 = 64;
+
+        let driver_source = VmTaskDriverSource::new(SingleDriverBackend::new(driver));
+        let base_len = 64 << 20;
+        let payload_len = 4 << 20;
+        let mem = DeviceSharedMemory::new(base_len, payload_len);
+        let mut msi_set = MsiInterruptSet::new();
+        let nvme = NvmeController::new(
+            &driver_source,
+            mem.guest_memory().clone(),
+            &mut msi_set,
+            &mut ExternallyManagedMmioIntercepts,
+            NvmeControllerCaps {
+                msix_count: MSIX_COUNT,
+                max_io_queues: IO_QUEUE_COUNT,
+                subsystem_id: Guid::new_random(),
+            },
+        );
+
+        let payload_mem = mem
+            .guest_memory()
+            .subrange(base_len as u64, payload_len as u64, false)
+            .unwrap();
+        let mut buf = vec![0u8; sector_count as usize * sector_size as usize];
+        let mut temp = vec![0u8; sector_size as usize];
+        assert!(sector_size > 2);
+        temp[sector_size as usize / 2 - 1] = 2;
+        temp[sector_size as usize / 2] = 3;
+
+        for i in (0..buf.len()).step_by(temp.len()) {
```

Comment:
You can use `chunks_exact_mut` to get an iterator of chunks that are each `temp.len()` length long.

Reply:
(and this looks like a duplication of the code in `make_repeat_data_buffer` ... did you intend to just use that function?